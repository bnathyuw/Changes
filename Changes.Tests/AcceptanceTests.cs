using System.Collections.Generic;
using System.Linq;
using ApprovalTests.Reporters;
using FluentAssertions;
using Xunit;
using static System.Environment;
using static System.String;
using static ApprovalTests.Approvals;
using static Changes.Domain.Transformations;
using Changes.Domain;

namespace Changes.Tests
{
    [UseReporter(typeof(XUnit2Reporter))]
    public class AcceptanceTests
    {
        private static readonly Transformations OnceOnlyTransformations = FromPlaceNotation("x");
        private static readonly Transformations PlainHuntTransformations = FromPlaceNotation("x1");
        private static readonly Transformations StNicholasBobTransformations = FromPlaceNotation("34.14x34,12");
        private static readonly Transformations StedmanDoublesTransformations = FromPlaceNotation("3.1.5.3.1.3,1");
        private static readonly Transformations StedmanDoublesWithSingle = FromPlaceNotation("3.1.5.3.1.345,1");
        private static readonly Transformations StedmanDoublesWithSingleInLastLead = 
            StedmanDoublesTransformations * 4 + StedmanDoublesWithSingle;

        [Fact]
        public void TwoElementSequence()
        {
            var changes = new Generator(new object[] {1, 2}, OnceOnlyTransformations)
                .ToArray();

            changes.Should().HaveCount(2);
            Verify(changes.ToText());
        }


        [Fact]
        public void ThreeElementSequence()
        {
            var changes = new Generator(new object[] {'a', 'b', 'c'}, PlainHuntTransformations)
                .ToArray();

            changes.Should().HaveCount(6);
            Verify(changes.ToText());
        }

        [Fact]
        public void StedmanDoubles()
        {
            var changes = new Generator(new object[] {1, 2, 3, 4, 5}, StedmanDoublesTransformations)
                .ToArray();

            changes.Should().HaveCount(60);
            Verify(changes.ToText());
        }

        [Fact]
        public void StedmanDoublesWithSingles()
        {
            var changes = new Generator(new object[] {1, 2, 3, 4, 5}, StedmanDoublesWithSingleInLastLead)
                .ToArray();

            changes.Should().HaveCount(120);
            Verify(changes.ToText());
        }

        [Fact]
        public void StNicholasBobMinimus()
        {
            var changes = new Generator(new object[] {1, 2, 3, 4, 5}, StNicholasBobTransformations).ToArray();

            changes.Should().HaveCount(24);
            Verify(changes.ToText());
        }

        [Fact]
        public void EmmaThomas()
        {
            var one = new Generator(new object[] {"mas"}, OnceOnlyTransformations).ToArray();
            var two = new Generator(new object[] {"Tho", "mas"}, OnceOnlyTransformations).ToArray();
            var three = new Generator(new object[] {"et", "Tho", "mas"}, PlainHuntTransformations).ToArray();
            var four = new Generator(new object[] {"ma", "et", "Tho", "mas"}, StNicholasBobTransformations).ToArray();
            var five = new Generator(new object[] {"Em", "ma", "et", "Tho", "mas"}, StedmanDoublesWithSingleInLastLead).ToArray();
            var six = new Generator(new object[] {"Em", "ma", "et", "Tho"}, StNicholasBobTransformations).ToArray();
            var seven = new Generator(new object[] {"Em", "ma", "et"}, PlainHuntTransformations).ToArray();
            var eight = new Generator(new object[] {"Em", "ma"}, OnceOnlyTransformations).ToArray();
            var nine = new Generator(new object[] {"Em"}, OnceOnlyTransformations).ToArray();

            Verify(FormatChanges.ToTabbedText(one, two, three, four, five, six, seven, eight, nine));
        }
    }

    public static class FormatChanges
    {
        public static string ToText(this IEnumerable<object[]> changes) => Join(NewLine, changes.Select(ToUndelimitedList));
        private static string ToUndelimitedList(object[] x) => Join("", x);
        public static string ToTabbedText(params object[][][] peals) => Join(NewLine, peals.Select(AsLines));
        private static string AsLines(object[][] changes) => Join(NewLine, changes.Select(TabbedList));
        private static string TabbedList(object[] x) => Join("\t", x);
    }
}