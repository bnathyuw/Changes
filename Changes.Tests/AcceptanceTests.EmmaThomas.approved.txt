mas
Tho	mas
mas	Tho
et	Tho	mas
Tho	et	mas
Tho	mas	et
mas	Tho	et
mas	et	Tho
et	mas	Tho
ma	et	Tho	mas
et	ma	Tho	mas
et	Tho	ma	mas
Tho	et	mas	ma
et	Tho	mas	ma
Tho	et	ma	mas
Tho	ma	et	mas
ma	Tho	et	mas
ma	Tho	mas	et
Tho	ma	mas	et
Tho	mas	ma	et
mas	Tho	et	ma
Tho	mas	et	ma
mas	Tho	ma	et
mas	ma	Tho	et
ma	mas	Tho	et
ma	mas	et	Tho
mas	ma	et	Tho
mas	et	ma	Tho
et	mas	Tho	ma
mas	et	Tho	ma
et	mas	ma	Tho
et	ma	mas	Tho
ma	et	mas	Tho
Em	ma	et	Tho	mas
ma	Em	et	mas	Tho
ma	et	Em	Tho	mas
et	ma	Tho	Em	mas
ma	et	Tho	mas	Em
ma	Tho	et	Em	mas
Tho	ma	et	mas	Em
Tho	et	ma	Em	mas
et	Tho	ma	mas	Em
Tho	et	mas	ma	Em
Tho	mas	et	Em	ma
mas	Tho	et	ma	Em
mas	et	Tho	Em	ma
et	mas	Tho	ma	Em
et	Tho	mas	Em	ma
Tho	et	Em	mas	ma
et	Tho	Em	ma	mas
et	Em	Tho	mas	ma
Em	et	Tho	ma	mas
Em	Tho	et	mas	ma
Tho	Em	et	ma	mas
Em	Tho	ma	et	mas
Em	ma	Tho	mas	et
ma	Em	Tho	et	mas
ma	Tho	Em	mas	et
Tho	ma	Em	et	mas
Tho	Em	ma	mas	et
Em	Tho	mas	ma	et
Tho	Em	mas	et	ma
Tho	mas	Em	ma	et
mas	Tho	Em	et	ma
mas	Em	Tho	ma	et
Em	mas	Tho	et	ma
mas	Em	et	Tho	ma
mas	et	Em	ma	Tho
et	mas	Em	Tho	ma
et	Em	mas	ma	Tho
Em	et	mas	Tho	ma
Em	mas	et	ma	Tho
mas	Em	ma	et	Tho
Em	mas	ma	Tho	et
Em	ma	mas	et	Tho
ma	Em	mas	Tho	et
ma	mas	Em	et	Tho
mas	ma	Em	Tho	et
ma	mas	Tho	Em	et
ma	Tho	mas	et	Em
Tho	ma	mas	Em	et
Tho	mas	ma	et	Em
mas	Tho	ma	Em	et
mas	ma	Tho	et	Em
ma	mas	et	Tho	Em
mas	ma	et	Em	Tho
mas	et	ma	Tho	Em
et	mas	ma	Tho	Em
et	ma	mas	Em	Tho
ma	et	mas	Tho	Em
et	ma	Tho	mas	Em
et	Tho	ma	Em	mas
Tho	et	ma	mas	Em
Tho	ma	et	Em	mas
ma	Tho	et	mas	Em
ma	et	Tho	Em	mas
et	ma	Em	Tho	mas
ma	et	Em	mas	Tho
ma	Em	et	Tho	mas
Em	ma	et	mas	Tho
Em	et	ma	Tho	mas
et	Em	ma	mas	Tho
Em	et	mas	ma	Tho
Em	mas	et	Tho	ma
mas	Em	et	ma	Tho
mas	et	Em	Tho	ma
et	mas	Em	ma	Tho
et	Em	mas	Tho	ma
Em	et	Tho	mas	ma
et	Em	Tho	ma	mas
et	Tho	Em	mas	ma
Tho	et	Em	ma	mas
Tho	Em	et	mas	ma
Em	Tho	et	ma	mas
Tho	Em	ma	et	mas
Tho	ma	Em	mas	et
ma	Tho	Em	et	mas
ma	Em	Tho	mas	et
Em	ma	Tho	et	mas
Em	Tho	ma	mas	et
Tho	Em	mas	ma	et
Em	Tho	mas	et	ma
Em	mas	Tho	ma	et
mas	Em	Tho	et	ma
mas	Tho	Em	ma	et
Tho	mas	Em	et	ma
mas	Tho	et	Em	ma
mas	et	Tho	ma	Em
et	mas	Tho	Em	ma
et	Tho	mas	ma	Em
Tho	et	mas	Em	ma
Tho	mas	et	ma	Em
mas	Tho	ma	et	Em
Tho	mas	ma	Em	et
Tho	ma	mas	et	Em
ma	Tho	mas	Em	et
ma	mas	Tho	et	Em
mas	ma	Tho	Em	et
ma	mas	Em	Tho	et
ma	Em	mas	et	Tho
Em	ma	mas	Tho	et
Em	mas	ma	et	Tho
mas	Em	ma	Tho	et
mas	ma	Em	et	Tho
ma	mas	et	Em	Tho
mas	ma	et	Tho	Em
mas	et	ma	Em	Tho
et	mas	ma	Em	Tho
et	ma	mas	Tho	Em
ma	et	mas	Em	Tho
et	ma	Em	mas	Tho
et	Em	ma	Tho	mas
Em	et	ma	mas	Tho
Em	ma	et	Tho
ma	Em	et	Tho
ma	et	Em	Tho
et	ma	Tho	Em
ma	et	Tho	Em
et	ma	Em	Tho
et	Em	ma	Tho
Em	et	ma	Tho
Em	et	Tho	ma
et	Em	Tho	ma
et	Tho	Em	ma
Tho	et	ma	Em
et	Tho	ma	Em
Tho	et	Em	ma
Tho	Em	et	ma
Em	Tho	et	ma
Em	Tho	ma	et
Tho	Em	ma	et
Tho	ma	Em	et
ma	Tho	et	Em
Tho	ma	et	Em
ma	Tho	Em	et
ma	Em	Tho	et
Em	ma	Tho	et
Em	ma	et
ma	Em	et
ma	et	Em
et	ma	Em
et	Em	ma
Em	et	ma
Em	ma
ma	Em
Em